### ISSUE TITLE
Add Option for Marginal Cost Offer Curve

### ISSUE THREAD

**Issue by tpblaster** [PyPSA/PyPSA#1020]
It would be extremely helpful to add multipart offer curves to generators, rather than just having a static marginal cost or quadratic marginal cost. In many balancing authorities, generators as well as loads are allowed to submit stepwise increasing offer curves rather than a single offer for their whole capacity. This feature would make PyPSA more robust and better able to capture unit economic dynamics. 

This feature would require a new variable that may or may not be time varying. This variable would be represented as a series that would break generator's marginal cost into chunks based on percentage of generation. For example, a 100mw generator might be represented in this manner:
20mw for $20/mw
40mw for $30/mw
35mw for $35/mw
5mw for $50/mw
This could be captured in a series of tuples formatted like ((.2, 20), (.4, 30), (.35, 35), (.05, 50))

In addition to the actual implementation, several consistency checks would need to be added. First would be a check to make sure that the total capacity being broken up summed to one. The marginal cost curve would also need to be monotonically increasing and may be limited to a set number of offers, ten is likely more than enough. 

Files that would need changes:
variables.csv
variables.py
consistency.py
optimize.py

It may also be prudent to change the logic in define_objective to include an enum to store the offer type for a generator, rather than using a boolean is_quadratic variable.

**Comment by tpblaster** [comment]
Here is the way to change the model in order to handle this:

For each offer we must create a new variable and a constraint to set whether that marginal cost should be used, this can be done in the following manner:

New Constraints
(offer lower bound) * p <= x1
x1 <= p * (offer upper bound)
sum(x1, xN...) = 1

Where p is a units generation for a given snapshot and x1 is a boolean variable.

Objective function:
mc1 * operation * x1 + mc2 * operation * x2
where mc1 and mc2 are marginal costs and x1 and x2 are defined as constraints like above. 

This can be extended to any number of offers, any feedback would be appreciated.

**Comment by fneum** [comment]
The recommended way for step-wise marginal cost curves in PyPSA is to add multiple generators where the `marginal_cost` represents the price of the step and `p_nom` represents the volume of the step's bid. Would that work for your application?

**Comment by tpblaster** [comment]
No, for my application I would like to be able to capture short run electricity market dispatch, which needs to take into account startup costs and other time varying parameters. This is also much less clunky than having multiple generators.

**Comment by liu-jinyu** [comment]
As a power market researcher, I really feel like this functionality is super important. This could enable more general and accurate power market simulations!!

**Comment by aliati94** [comment]
Totally agree with this. Breaking each generator's offer curve into multiple generators makes it so much more complicated.

**Comment by fneum** [comment]
Ranking this a bit higher in the priority list of new features due to popular demand for it here and bilaterally.

**Comment by fneum** [comment]
Could be implemented potentially with SOS2 constraints (when first bids are higher):

https://github.com/PyPSA/linopy/issues/437

**Comment by joechen1579** [comment]
> I implemented a version that supports 10 segment bids: https://github.com/liu-jinyu/PyPSA/tree/v0.31.667
> 
> You can use it by setting a generator with use_bid_curve=True, and with its bid_priceX and bid_powerX properties are properly set.
> 
> Note the bid_priceX should be monotonically increasing, and bid_powerX should be positive power segments. Sum over bid_powerX is the total bid power.
> 
> Suggestions for this bidding curve implementation are welcomed!

Hi Liu, thanks for sharing this repo!

I’m working on a UC model in PyPSA, and your bid_curve implementation is really helpful on what I am developing. Would you have an example of how to apply bid_curve to generators? I tried a simple UC with two generators using bid_curve, but I’m not getting the expected dispatch results. I am also more than happy to provide my simple UC model for discussion.
